### PR TITLE
Stats calculator fixes [Part 2]

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 1]
```diff
@@ -353,38 +353,38 @@ else if (leftNDV <= matchingRightNDV) {
 
     @VisibleForTesting
     PlanNodeStatsEstimate addJoinComplementStats(
-            PlanNodeStatsEstimate sourceStats,
-            PlanNodeStatsEstimate baseJoinStats,
+            PlanNodeStatsEstimate leftStats,
+            PlanNodeStatsEstimate innerJoinStats,
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Continuing the discussion from https://github.com/prestodb/presto/pull/11856#discussion_r232167964

I like the old names better. Do you feel strongly about this rename?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 2]
```diff
@@ -243,10 +244,12 @@ private PlanNodeStatsEstimate filterByAuxiliaryClause(PlanNodeStatsEstimate stat
...
                 .build();
 
-        return stats
-                .mapSymbolColumnStatistics(clause.getLeft(), oldLeftStats -> newLeftStats)
-                .mapSymbolColumnStatistics(clause.getRight(), oldRightStats -> newRightStats)
-                .mapOutputRowCount(rowCount -> rowCount * UNKNOWN_FILTER_COEFFICIENT);
+        PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.buildFrom(stats);
+        result.setOutputRowCount(stats.getOutputRowCount() * UNKNOWN_FILTER_COEFFICIENT);
+        result.addSymbolStatistics(clause.getLeft(), newLeftStats);
+        result.addSymbolStatistics(clause.getRight(), newRightStats);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: continuing https://github.com/prestodb/presto/pull/11856#discussion_r232168764

you still can use builder fluently like below; this will be more readable IMO
```
PlanNodeStatsEstimate.Builder result = PlanNodeStatsEstimate.buildFrom(stats)
    .setOutputRowCount(stats.getOutputRowCount() * UNKNOWN_FILTER_COEFFICIENT)
    .addSymbolStatistics(cla ...

return normalizer.normalize(result.build(), types);
```

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 3]
```diff
@@ -273,7 +273,7 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
             PlanNodeStatsEstimate rightStats,
             TypeProvider types)
     {
-        if (rightStats.getOutputRowCount() == 0) {
+        if (leftStats.getOutputRowCount() == 0 || rightStats.getOutputRowCount() == 0) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Can we have a test for that? Especially in one of the `TestTpc*Stats` classes?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 4]
```diff
@@ -358,8 +358,12 @@ PlanNodeStatsEstimate addJoinComplementStats(
             PlanNodeStatsEstimate innerJoinStats,
             PlanNodeStatsEstimate joinComplementStats)
     {
-        double innerJoinRowCount = innerJoinStats.getOutputRowCount();
         double joinComplementRowCount = joinComplementStats.getOutputRowCount();
+        if (joinComplementRowCount == 0) {
+            return innerJoinStats;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Can we have a test for that? Especially in one of the `TestTpc*Stats` classes?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 5]
```diff
@@ -65,6 +65,7 @@ public JoinStatsRule(FilterStatsCalculator filterStatsCalculator, StatsNormalize
     {
         super(normalizer);
         this.filterStatsCalculator = requireNonNull(filterStatsCalculator, "filterStatsCalculator is null");
+        this.normalizer = normalizer;
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: It cannot assess the impact of this change. I know there are tests and the tests have their expectations appropriately updated, but these are synthetic tests and are as good as they expected values.

Would it be possible to bring more confidence to this change by adding a test in one of the `TestTpc*Stats` classes?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 6]
```diff
@@ -15,7 +15,6 @@
 
 import com.facebook.presto.Session;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Discussion:

> When doing stats arithmetics we expect that the symbol stats are normalized. I didn't catch an explicit failure there, but it looks like dealing on normalizing stats is safer.

Normalizing can also increase error and introduce noise. For instance you would expect join tree stats to be the same at root independently on join order. However, this is not the case due to normalization. Let have two joins, `J1, J2`.
Let's say filter factor for join `J1` is `1/10` and for `J2` `10` (multiplies rows).

If on input you have 10 rows with 10 distinct values (on join key `K`) then if you first join `J1` and then `J2` then for column `K` you will end up with just 1 distinct value (normalization after `J1`).
If you first use `J2` and then `J1` then `K` will still have 10 distinct values.

Therefore I would drop the change if it's not motivated by real use case.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 2 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Oh, that's what you meant. Agreed.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 6 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: `JoinStatsRule` rule is a `SimpleStatsRule`, that does the normalization of the final stats anyway. Therefore when estimating multiple joins, the intermediate stats are normalized anyway.

This patch only adds normalization to the intermediate stats, and ensure that the stats are normalized before applying the `addJoinComplementStats` logic. This follows the logic we use in estimating filters. The stats are normalized at every step.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 3 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Since we are not going to round, explicit check `== 0` is misleading. Given that there are no divisions or multiplications by the rows count it looks safe to just remove this special case check altogether.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Could you please elaborate more on this?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 3 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: I'm reverting this change. As there's no division or multiplication involving `leftStats.getOutputRowCount()`

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: We have many tests for JOIN in `TestTpc*Stats`. What additional tests would you like to see?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 4 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: `TestTpchLocalStats#testFullJoin`, `TestTpchLocalStats#testRightJoin`, `TestTpchLocalStats#testLeftJoin` tests cover this branch

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 5 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Hm.. but none of these needed (nor enjoyed) an update. So there is nothing ensuring we don't regress in the future.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: as in https://github.com/prestodb/presto/pull/11895#discussion_r233997741

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Tests in `TestTpc*Stats` are not checking for exact estimate, and allow deviations.

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: I know that. If the change improves the estimations, "allowed deviations" could be reduced.
If this is not possible (or does not make sense; or anyhow would not ensure we do not regress), then we need an explicit test.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: I'm not sure if normalization improves accuracy. That's not the point of this commit. JoinStatsRule is logic intensive. A lot of arithmetic is being done. The FilterStatsCalculator is used in multiple places. As everywhere else across the code (including `FilterStatsCalculator`) it is assumed that stats are normalized. (e.g. rowCount >= ndv, ndv consistent with range, nulls factor consistent with ndv, etc.) it seems to be just safer to do not break this assumption in the `JoinStatsRule` as well.

Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Fair enough. Keep in mind that join stats rule is executed in tight loop within join enumeration so it has to be performant therefore we should use normalization only where it has chance to fix something. 

@findepi do you recall when normalization made plans worse?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: > @findepi do you recall if/when normalization made plans worse?

unfortunately i don't remember. It was introduced quite some time ago, to avoid totally inconsistent stats (eg. NDV > row count after a filter for symbols not involved in the filter condition), as they made further rules produce stupid results (like `x = c` producing < 1 rows, even though we assume this should produce 1+).

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 7]
```diff
@@ -280,24 +283,21 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
...
 
-            return leftStats.mapOutputRowCount(rowCount -> 0.0);
+            return normalizer.normalize(leftStats.mapOutputRowCount(rowCount -> 0.0), types);
         }
 
         // TODO: add support for non-equality conditions (e.g: <=, !=, >)
-        int numberOfFilterClauses = filter.map(
-                exression -> ExpressionUtils.extractConjuncts(exression).size())
-                .orElse(0);
+        int numberOfFilterClauses = filter.map(expression -> extractConjuncts(expression).size()).orElse(0);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: nit: this could be part of refactoring commit

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java [hunk 8]
```diff
@@ -280,24 +283,21 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
...
         // Principals behind this heuristics is the same as in computeInnerJoinStats:
         // select "driving join clause" that reduces matched rows the most.
-        return IntStream.range(0, criteria.size())
-                .mapToObj(drivingClauseId -> {
-                    EquiJoinClause drivingClause = criteria.get(drivingClauseId);
-                    return calculateJoinComplementStats(leftStats, rightStats, drivingClause, criteria.size() - 1 + numberOfFilterClauses);
-                })
+        return criteria.stream()
+                .map(drivingClause -> calculateJoinComplementStats(leftStats, rightStats, drivingClause, criteria.size() - 1 + numberOfFilterClauses))
+                .filter(estimate -> !estimate.isOutputRowCountUnknown())
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: This seems unrelated to normalizer introduction. Can't normalizer consume unknown stats (I think it should)?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 4 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: There is no division `joinComplementRowCount` anywhere. Should the check be: `outputRowCount == 0` instead?

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 8 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Let me extract this into a separate commit. This change favors known stats over unknown stats. Before if any of the stats were unknown, the result was unknown.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 4 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: I had that like this before. But in the previous PR you suggested to shortcircut if complement stats are 0. Correct me if i got you wrong.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 5 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: @sopel39 This is a very valid concert. I'm running the benchmarks to estimate the impact.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: Benchmarks didn't show any significant (visible) regressions after applying normalization.

Before:

```
# Run complete. Total time: 00:02:18

Benchmark                                           (joinReorderingStrategy)  Mode  Cnt    Score   Error  Units
BenchmarkReorderChainedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS  avgt   30   19.224 ± 0.626  ms/op
BenchmarkReorderChainedJoins.benchmarkReorderJoins                 AUTOMATIC  avgt   30  215.899 ± 4.233  ms/op

Benchmark                                                  (joinReorderingStrategy)  (numberOfTables)  Mode  Cnt      Score     Error  Units
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 2  avgt   30      6.109 ±   0.315  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 4  avgt   30      9.982 ±   0.478  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 6  avgt   30     13.560 ±   0.381  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 8  avgt   30     17.579 ±   0.425  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                10  avgt   30     22.274 ±   0.537  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 2  avgt   30      7.378 ±   0.442  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 4  avgt   30     19.342 ±   0.675  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 6  avgt   30    116.208 ±   4.158  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 8  avgt   30   1056.741 ±  28.317  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                10  avgt   30  10150.088 ± 149.229  ms/op

```

After:

```
Benchmark                                           (joinReorderingStrategy)  Mode  Cnt    Score   Error  Units
BenchmarkReorderChainedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS  avgt   30   19.305 ± 0.552  ms/op
BenchmarkReorderChainedJoins.benchmarkReorderJoins                 AUTOMATIC  avgt   30  217.149 ± 5.105  ms/op

# Run complete. Total time: 00:21:16

Benchmark                                                  (joinReorderingStrategy)  (numberOfTables)  Mode  Cnt      Score      Error  Units
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 2  avgt   30      6.221 ±    0.320  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 4  avgt   30      9.663 ±    0.375  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 6  avgt   30     13.216 ±    0.423  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                 8  avgt   30     17.506 ±    0.461  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins     ELIMINATE_CROSS_JOINS                10  avgt   30     22.591 ±    0.823  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 2  avgt   30      7.491 ±    0.554  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 4  avgt   30     20.584 ±    0.829  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 6  avgt   30    116.934 ±    3.790  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                 8  avgt   30   1105.047 ±   28.931  ms/op
BenchmarkReorderInterconnectedJoins.benchmarkReorderJoins                 AUTOMATIC                10  avgt   30  10309.552 ±  153.314  ms/op
```

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: It actually makes sense. The normalization is very cheap if it does nothing.

File: presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java (hunk 4 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java]: It is not in the loop. It is at the beginning of the `addJoinComplementStats`.
